### ISSUE TITLE
Linking a dynamic library product with --static-swift-stdlib fails due to missing DispatchStubs

### ISSUE THREAD

**Issue by ephemer** [finagolfin/swift-android-sdk#194]
Hi @finagolfin, thanks a lot for implementing and building this as a Swift SDK. This is a big step forward!

I noticed when testing this just now that `swift build --static-swift-stdlib` appears to fail because `DispatchStubs` cannot be found. This library is linked from static/dispatch/module.modulemap in the case of a static stdlib build. And indeed, the DispatchStubs.a static library does exist. It's just that the linker appears to be searching for it in the wrong location. If I add `swift build ... -Xlinker --verbose` I can see that ldd is trying to link the `.so` dependencies (as well?). So there appears to be some internal confusion here.

For now, we cannot remove our dependency on `Dispatch` entirely, so I will continue to try without `--static-swift-stdlib`. But I would be interested to hear if you have an idea of why this is happening. Again, we're trying to build a `.dynamic` library product via `swift build`, but are hoping that its dependencies (including the swift stdlib) will be linked statically.

**Comment by finagolfin** [comment]
I'll take a look, thanks for reporting. I'm trying to get that flag working better with Swift 6, but there may still be some holes.

**Comment by finagolfin** [comment]
Alright, finally dug into this, appears to be an upstream bug with building dynamic libraries against the static SDK libraries. I had tested this SDK bundle with the static stdlib and building executable products, which worked fine, but not to build Swift shared library products.

You can work around the problem with these steps:

1. Find the path where your static SDK libraries for your required Android build triple are installed with this command:
```
> swift sdk configure --show-configuration swift-6.0.3-RELEASE-android-24-0.1 aarch64-unknown-linux-android24
sdkRootPath: /home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/android-27c-sysroot
swiftResourcesPath: /home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/android-27c-sysroot/usr/lib/swift
swiftStaticResourcesPath: /home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/android-27c-sysroot/usr/lib/swift_static-aarch64
includeSearchPaths: not set
librarySearchPaths: not set
toolsetPaths: ["/home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/swift-toolset.json"]
```
2. Use the `swiftStaticResourcesPath` shown when building with SwiftPM:
```
> swift build --swift-sdk aarch64-unknown-linux-android24 --static-swift-stdlib -Xswiftc -L/home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/android-27c-sysroot/usr/lib/swift_static-aarch64/android
```
Note the slight modification made to that resource path, adding the `android` directory at the end. That got a dynamic library cross-compiled for me. You may want to use [the 6.0.3 SDK bundle from yesterday's CI run](https://github.com/finagolfin/swift-android-sdk/actions/runs/12629675448/attempts/1) too.

I will file an issue for this upstream, meanwhile, let me know if this workaround fixes it for now.

**Comment by marcprux** [comment]
Fabulous! I haven't tested the output yet, but is seems to build on macOS:

```
zap swift-everywhere/example-package-deckofplayingcards % ${HOME}/Library/Developer/Toolchains/swift-6.0.2-RELEASE.xctoolchain/usr/bin/swift build -c release --swift-sdk aarch64-unknown-linux-android24 --static-swift-stdlib -Xswiftc -L${HOME}/.swiftpm/swift-sdks/swift-6.0.2-RELEASE-android-24-0.1.artifactbundle/swift-6.0.2-release-android-24-sdk/android-27c-sysroot/usr/lib/swift_static-aarch64/android
Building for production...
[7/7] Linking libDeckOfPlayingCards.so
Build complete! (1.76s)

zap swift-everywhere/example-package-deckofplayingcards % du -sk .build/release/libDeckOfPlayingCards.so
9476	.build/release/libDeckOfPlayingCards.so

zap swift-everywhere/example-package-deckofplayingcards % patchelf --print-needed .build/release/libDeckOfPlayingCards.so
libdl.so
libc.so
```

versus:

```
zap swift-everywhere/example-package-deckofplayingcards % ~/Library/Developer/Toolchains/swift-6.0.2-RELEASE.xctoolchain/usr/bin/swift build -c release --swift-sdk aarch64-unknown-linux-android24
Building for production...
[7/7] Linking libDeckOfPlayingCards.so
Build complete! (1.27s)

zap swift-everywhere/example-package-deckofplayingcards % du -sk .build/release/libDeckOfPlayingCards.so
108	.build/release/libDeckOfPlayingCards.so

zap swift-everywhere/example-package-deckofplayingcards % patchelf --print-needed .build/release/libDeckOfPlayingCards.so
libswiftCore.so
libswift_Concurrency.so
libswift_StringProcessing.so
libswift_RegexParser.so
libdl.so
libc.so
```

I assume the `-L` flag will need to be added manually for the foreseeable future. I was hopeful that I could add the path to the `swiftStaticResourcesPath` or `librarySearchPaths` properties of the `swift-sdk.json` file, but neither seemed to work.

**Comment by finagolfin** [comment]
> I was hopeful that I could add the path to the swiftStaticResourcesPath or librarySearchPaths properties of the swift-sdk.json file, but neither seemed to work.

Yeah, modifying `swiftStaticResourcesPath` won't work because SwiftPM appears to ignore that for dynamic library products. I'm unsure about why setting `librarySearchPaths` wouldn't work either, didn't try that.

Thanks for validating the fix on macOS. 👍

**Comment by finagolfin** [comment]
I've filed an upstream issue, which you can see linked above. Adding the following flags on the command-line instead, `-Xswiftc -static-stdlib -Xswiftc -resource-dir -Xswiftc /home/finagolfin/.swiftpm/swift-sdks/swift-6.0.3-RELEASE-android-24-0.1.artifactbundle/swift-6.0.3-release-android-24-sdk/android-27c-sysroot/usr/lib/swift_static-aarch64`, appears to fix this issue also.
